File: MATLAB version/Visualization.m
function [  ] = Visualization( data, clustering_result, is_showingnoise )
%Input:
%     data:is the data to be analyzed
%     clustering_result:is the result of 'Extreme_Clustering.m'
%     is_showingnoise:is 'true' or 'false', if 'true', show noises
%Output: is a visualization of clustering result
sum=size(data,1);
dim=size(data,2);
if dim>2
    dis=pdist(data);
    dis=squareform(dis);
    [v,stress]=mdscale(dis,2,'criterion','strain');
    data=v;
end
a=['r','g','b','m','c'];
b=['o','+','*','x'];
for ii=1:max(clustering_result)
   h=0;
   z=zeros(sum,2);
   for jj=1:sum
      if(clustering_result(jj)==ii)
         h=h+1;
         z(h,:)=data(jj,:);
      end
   end
   z(h+1:sum,:)=[];
   s=mod(ii,5)+1;
   t=mod(ii,4)+1;
   scatter(z(:,1),z(:,2),a(s),b(t))
   set(gca,'FontSize',15);
   xlabel('attribute 1','fontsize',18);
   ylabel('attribute 2','fontsize',18);
   title('Extreme clustering');
   hold on
end
if is_showingnoise
    h=0;
    z=zeros(sum,2);
    for jj=1:sum
        if(clustering_result(jj)==-1)
            h=h+1;
            z(h,:)=data(jj,:);
        end
    end
    z(h+1:sum,:)=[];
    scatter(z(:,1),z(:,2),'k','.')
end

end

File: Python version/Visualization.py
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA


def Visualization(data, clusteringResult, isShowNoise=False):
    number = data.shape[0] 
    color = ['#278EA5','#21E6C1','#FFBA5A','#FF7657','#C56868','#6b76ff','#ff0000','#9e579d','#f85f73'
    ,'#928a97','#b6f7c1','#0b409c','#d22780','#882042','#071E3D']
    lineForm = ['o','+','*','x']
    if data.shape[1]>2:
        pca = PCA(n_components=2)
        data = pca.fit_transform(data)
    for i in range(np.max(clusteringResult)):
        Together = []
        flag = 0
        for j in range(number):
            if clusteringResult[j] == i+1:
                flag += 1
                Together.append(data[j])
        Together = np.array(Together)
        colorNum = np.mod(i+1,15)
        formNum = np.mod(i+1,4)
        plt.scatter(Together[:,0], Together[:,1], 15, color[colorNum], lineForm[formNum])
    
    plt.xlabel('attribute 1', fontsize=20)
    plt.ylabel('attribute 2', fontsize=20)
    plt.title('Extreme clustering')
    if isShowNoise == True:
        Together = []
        flag = 0
        for j in range(number):
            if clusteringResult[j] == -1:
                flag += 1
                Together.append(data[j])
        Together = np.array(Together, dtype = int)
        if Together.shape[0] != 0:
            plt.scatter(Together[:,0], Together[:,1], 15, 'k', '.')
    plt.show()